### PR TITLE
[Dynamic Sampling] Head of trace

### LINKED ISSUE #457: feat(tracing) Collect DSC for head of trace
### **Describe the idea**
Applications instrumented with a Sentry SDK should collect and propagate dynamic sampling context data to all outgoing requests and to Sentry.

### Requirement:
**When** an application has an event which is the head of a trace,
**the SDK shall** collect and propagate all possible values in outgoing requests and to Sentry 
**so that** contextual data can be propagated in all connected services across the continuation of trace

_what should be sent?_

[baggage header](https://develop-b38lax7t3.sentry.dev/sdk/performance/dynamic-sampling-context/#baggage-header)

- [ ] sentry-trace_id
- [ ] sentry-public_key
- [ ] sentry-sample_rate
- [ ] sentry-release
- [ ] sentry-environment
- [ ] sentry-user_id
- [ ] sentry-user_segment
- [ ] sentry-transaction

[Trace Envelope Header](https://develop-b38lax7t3.sentry.dev/sdk/performance/dynamic-sampling-context/#envelope-header)

- [ ] trace_id (string) 
- [ ] public_key (string)
- [ ] sample_rate (string) 
- [ ] release (string) 
- [ ] environment (string) 
- [ ] user_id (string) 
- [ ] user_segment (string) 
- [ ] transaction (string) 

_where should it be sent?_

- in dynamic sample context (by baggage)
- in the envelope header
- meta data (was/is being done in JS to be check if still necessary)

_Example:_
Here is an example from the Sentry UI, where the implementation has already begun on the latest versions of the JS SDK. A request from the front end to a down stream applications includes in the request header `baggage`

image

### **Why do you think it's beneficial to most of the users**
This ensures that further downstream applications connected in a distributed trace will receive all contextual data required for dynamically sampling.

### **Possible implementation**
Implementation has already begun (at the time of this creation) on JS and Android. Refer to the SDK Developer documentation for further information

For full up to date requirements according to SDK Developer specifications please see the associated documentation.

[Dynamic Sampling Context](https://develop-b38lax7t3.sentry.dev/sdk/performance/dynamic-sampling-context/)

Original PR for full context
Add spec for Dynamic Sampling Context #613

### PR DESCRIPTION
**Pull request by cleptric**
This adds support in case we're the head SDK that starts the trace for Dynamic Sampling.

- Add `DynamicSamplingContextFromTransaction`
- Add `Span.Source`
- Set the correct sample rate on the Span

Populating the envelop header will be done in #491 

Closes #457

### CODE REVIEW

File: dynamic_sampling_context.go [hunk 1]
```diff
@@ -16,7 +17,7 @@ type DynamicSamplingContext struct {
 	Frozen  bool
 }
 
-func NewDynamicSamplingContext(header []byte) (DynamicSamplingContext, error) {
+func NewDynamicSamplingContextFromHeader(header []byte) (DynamicSamplingContext, error) {
```
Comment (kamilogorek) [inline dynamic_sampling_context.go:20]: I'd use `DynamicSamplingContextFromHeader` instead. `NewX` should be reserved for direct builders.

File: dynamic_sampling_context.go [hunk 2]
```diff
@@ -36,10 +37,52 @@ func NewDynamicSamplingContext(header []byte) (DynamicSamplingContext, error) {
...
+	scope := hubFromContext(span.Context()).Scope()
+
+	if traceID := span.TraceID.String(); traceID != "" {
+		entries["trace_id"] = traceID
+	}
+	if sampleRate := span.sampleRate; sampleRate != 0 {
+		entries["sample_rate"] = strconv.FormatFloat(sampleRate, 'E', -1, 64)
+	}
+
+	hub := hubFromContext(span.Context())
```
Comment (kamilogorek) [inline dynamic_sampling_context.go:51]: You already asked for hub 10 lines above. You can reuse it.

File: dynamic_sampling_context.go [hunk 3]
```diff
@@ -36,10 +37,52 @@ func NewDynamicSamplingContext(header []byte) (DynamicSamplingContext, error) {
...
+
+	if dsn := hub.Client().dsn; dsn != nil {
+		if publicKey := dsn.publicKey; publicKey != "" {
+			entries["public_key"] = publicKey
+		}
+	}
+	if release := hub.Client().Options().Release; release != "" {
```
Comment (kamilogorek) [inline dynamic_sampling_context.go:58]: Would extract it to `clientOptions` for readability

**Review by kamilogorek — COMMENTED** [review]
Small nitpicks

File: dynamic_sampling_context.go [hunk 4]
```diff
@@ -36,10 +37,57 @@ func NewDynamicSamplingContext(header []byte) (DynamicSamplingContext, error) {
...
+	options := hub.Client().Options()
+
+	if traceID := span.TraceID.String(); traceID != "" {
+		entries["trace_id"] = traceID
+	}
+	if sampleRate := span.sampleRate; sampleRate != 0 {
+		entries["sample_rate"] = strconv.FormatFloat(sampleRate, 'f', -1, 64)
+	}
+
+	if dsn := hub.Client().dsn; dsn != nil {
```
Comment (AbhiPrasad) [inline dynamic_sampling_context.go:54]: l: can we declare `client := hub.Client()` above and re-use it for both the options call and the dsn call?

File: dynamic_sampling_context.go [hunk 5]
```diff
@@ -36,10 +37,57 @@ func NewDynamicSamplingContext(header []byte) (DynamicSamplingContext, error) {
...
+		entries["release"] = release
+	}
+	if environment := options.Environment; environment != "" {
+		entries["environment"] = environment
+	}
+
+	// Only include the transaction name if it's of good quality (not empty and not SourceURL)
+	if span.Source != "" && span.Source != SourceURL {
+		if transactioName := scope.Transaction(); transactioName != "" {
+			entries["transaction"] = transactioName
```
Comment (AbhiPrasad) [inline dynamic_sampling_context.go:69]:
```suggestion
		if transactionName := scope.Transaction(); transactionName != "" {
			entries["transaction"] = transactionName
```

File: tracing.go [hunk 6]
```diff
@@ -459,6 +481,18 @@ var (
 	zeroSpanID  SpanID
 )
 
+// Contains information about how the name of the transaction was determined.
+type Source string
```
Comment (AbhiPrasad) [inline tracing.go:485]: m: I would like to call this `TransactionSource` (and rename the enum below as well) to make it super clear to the reader - `Source` is kind of ambiguous.

File: dynamic_sampling_context.go [hunk 7]
```diff
@@ -36,10 +37,57 @@ func NewDynamicSamplingContext(header []byte) (DynamicSamplingContext, error) {
...
+		}
+	}
+
+	if userSegment := scope.user.Segment; userSegment != "" {
+		entries["user_segment"] = userSegment
```
Comment (AbhiPrasad) [inline dynamic_sampling_context.go:74]: m: should we gate this by `sendDefaultPII`?

File: dynamic_sampling_context.go [hunk 8]
```diff
@@ -36,10 +37,57 @@ func NewDynamicSamplingContext(header []byte) (DynamicSamplingContext, error) {
 	}, nil
 }
 
+func DynamicSamplingContextFromTransaction(span *Span) (DynamicSamplingContext, error) {
```
Comment (AbhiPrasad) [inline dynamic_sampling_context.go:40]: m: Should this return an error?

File: http/sentryhttp.go [hunk 9]
```diff
@@ -71,9 +71,9 @@ func (h *Handler) Handle(handler http.Handler) http.Handler {
 // where that is convenient. In particular, use it to wrap a handler function
 // literal.
 //
-//  http.Handle(pattern, h.HandleFunc(func (w http.ResponseWriter, r *http.Request) {
-//      // handler code here
-//  }))
+//	http.Handle(pattern, h.HandleFunc(func (w http.ResponseWriter, r *http.Request) {
+//	    // handler code here
+//	}))
```
Comment (AbhiPrasad) [inline http/sentryhttp.go:76]: why did this add spaces here?

File: dynamic_sampling_context.go (hunk 7 above)
Comment (cleptric) [reply dynamic_sampling_context.go:74]: Not sure, did we do this in other SDKs?
In the end, this is explicitly set by the user.

File: tracing.go (hunk 6 above)
Comment (cleptric) [reply tracing.go]: https://github.com/getsentry/sentry-go/pull/492/commits/204e87b8bbb7c448b266ab7987d1d370ebdbeb4b

File: dynamic_sampling_context.go (hunk 4 above)
Comment (cleptric) [reply dynamic_sampling_context.go]: https://github.com/getsentry/sentry-go/pull/492/commits/b182fc54e4fa80621648d2a6b006b359593b135b

File: dynamic_sampling_context.go (hunk 8 above)
Comment (cleptric) [reply dynamic_sampling_context.go]: https://github.com/getsentry/sentry-go/pull/492/commits/b182fc54e4fa80621648d2a6b006b359593b135b

File: http/sentryhttp.go (hunk 9 above)
Comment (cleptric) [reply http/sentryhttp.go:76]: Happens on save, some `go fmt` stuff

File: dynamic_sampling_context.go (hunk 7 above)
Comment (AbhiPrasad) [reply dynamic_sampling_context.go:75]: You're right, let's keep this then.

**Review by AbhiPrasad — APPROVED** [review]
:shipit:

**Pull request merged**